### PR TITLE
Add S3 group generics to `.base_s3_generics`, include exported S3 generics in generic list

### LINKED ISSUE #1841: `object_name_linter()` not detecting all generics
``` r
lintr::lint(text = "`==.snake_case` <- function(a, b) a + b")
#> <text>:1:1: style: [object_name_linter] Variable and function name style should match snake_case or symbols.
#> `==.snake_case` <- function(a, b) a + b
#> ^~~~~~~~~~~~~~~
```

<sup>Created on 2022-12-13 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by AshesITR**
fixes #1841 and #1808

### CODE REVIEW

File: R/namespace.R [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Parse namespace files and return imports exports, methods
-namespace_imports <- function(path = find_package()) {
+namespace_imports <- function(path = find_package(".")) {
```
Comment (AshesITR) [inline R/namespace.R:2]: The previous default was an error.

File: R/namespace.R [hunk 2]
```diff
@@ -35,7 +35,7 @@ safe_get_exports <- function(ns) {
 }
 
 empty_namespace_data <- function() {
-  data.frame(pkg = character(), ns = character(), stringsAsFactors = FALSE)
+  data.frame(pkg = character(), fun = character(), stringsAsFactors = FALSE)
```
Comment (AshesITR) [inline R/namespace.R:38]: All other code paths return a data frame with column names `pkg` and `fun`.

File: tests/testthat/dummy_packages/clean/R/clean_generics.R [hunk 3]
```diff
@@ -37,3 +37,10 @@ head.my_s3_object <- function(x, ...) {
 `names<-.my_custom_class` <- function(x, value) {
   NULL
 }
+
+#' @export
+#' Defined S3 generic in R/eat_me.R
+#' Tests #1808
+eat_me.liiiiiiiiiiiiiiiiiiiiiiiiiiist <- function(x, ...) {
```
Comment (AshesITR) [inline tests/testthat/dummy_packages/clean/R/clean_generics.R:44]: Class name has length 30, so this provokes `object_length_linter()`.

File: tests/testthat/dummy_packages/clean/NAMESPACE [hunk 4]
```diff
@@ -1 +1,10 @@
-importFrom(utils,head)
+# Generated by roxygen2: do not edit by hand
+
+S3method("names<-",my_custom_class)
+S3method(drink_me,data.frame)
+S3method(drink_me,default)
+S3method(drink_me,list)
+S3method(eat_me,liiiiiiiiiiiiiiiiiiiiiiiiiiist)
+S3method(head,my_s3_object)
```
Comment (AshesITR) [inline tests/testthat/dummy_packages/clean/NAMESPACE:8]: These are necessary for fixing #1808

**Comment by AshesITR** [comment]
I'm looking into it, but it's slow for me to debug on old r versions.

**Comment by AshesITR** [comment]
🤦🏼‍♂️ the old `stringsAsFactors = TRUE` bit me.
image

File: R/namespace.R [hunk 5]
```diff
@@ -75,5 +92,16 @@ is_s3_generic <- function(fun) {
     # R < 3.5.0 doesn't provide .S3_methods_table
     # fallback: search baseenv() for generic methods
     imported_s3_generics(data.frame(pkg = "base", fun = ls(baseenv()), stringsAsFactors = FALSE))$fun
-  }
-)
+  },
+  # S3 generic groups, see ?base::groupGeneric
+  # Group "Math"
+  c("abs", "sign", "sqrt", "floor", "ceiling", "trunc", "round", "signif", "exp", "log", "expm1", "log1p", "cos",
```
Comment (MichaelChirico) [inline R/namespace.R:98]: Missing `log2()` and `log10()` right?

https://github.com/r-devel/r-svn/blob/bb4492e2a913a187ae77fe1f81329a2c81c320a7/src/library/base/R/zzz.R#L178-L184

Comment (MichaelChirico) [reply R/namespace.R:98]: WDYT about using `ls(.GenericArgsEnv)` instead? That would also subsume `.S3PrimitiveGenerics`.

Comment (AshesITR) [reply R/namespace.R:98]: Sounds good. I'll try this and see if all R versions work with that.

Comment (AshesITR) [reply R/namespace.R]: Update: Looks like it works. Thanks for the tip!

**Review by MichaelChirico — APPROVED** [review]

**Pull request merged**